### PR TITLE
Add a cache on top of python27 hashing

### PR DESCRIPTION
**Pull request by acristin**
Profiling indicated that we spend a lot of time in our custom deepcopy, mostly due to maintaining compatibility with python2.7 hashes and their ordering of keys. If we change the ordering of keys, we will change logical IDs of some generated resources, breaking backwards compatibility.

We found that during the hashing code, we made some 13MM calls to Hash.hash, but with caching this is reduced to 1040 calls for a large test template.

On my laptop, this reduces wall clock time from ~15s to ~13s, or ~34s to ~27s with cProfiler.

### Checklist

- [ ] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [ ] Add/update [transform tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions)
    - [ ] Using correct values
    - [ ] Using wrong values
- [ ] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)

### Examples?

Please reach out in the comments if you want to add an example. Examples will be 
added to `sam init` through [aws/aws-sam-cli-app-templates](https://github.com/aws/aws-sam-cli-app-templates).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by aahung — APPROVED** [review]

**Comment by aahung** [comment]
on 2019 16-inch macbook pro, 42.8s -> 34.8s: 19% less time

File: samtranslator/third_party/py27hash/hash.py [hunk 1]
```diff
@@ -28,6 +42,7 @@ class Hash(object):
     """
 
     @staticmethod
+    @lru_cache(maxsize=2048)
```
Comment (aahung) [inline samtranslator/third_party/py27hash/hash.py:45]: we should move the "str" check to the top as that's what hits the most

File: samtranslator/third_party/py27hash/hash.py [hunk 2]
```diff
@@ -2,11 +2,25 @@
...
+from typing import TYPE_CHECKING, TypeVar, Callable, Any
+
+if TYPE_CHECKING:
+    F = TypeVar("F", bound=Callable[[Any], Any])
+
+    def lru_cache(maxsize: int = 128, typed: bool = False) -> Callable[[F], F]:
+        pass
+
+
+else:
```
Comment (hoffa) [inline samtranslator/third_party/py27hash/hash.py:21]: Any benefit versus `# type: ignoring` with a comment?

Comment (acristin) [reply samtranslator/third_party/py27hash/hash.py:21]: I was probably doing it wrong, but I was unable to make `# type:ignoring` work; I still got complaints about unused type annotations as in original commit.

File: samtranslator/third_party/py27hash/hash.py [hunk 3]
```diff
@@ -2,11 +2,25 @@
 Compatibility methods to support Python 2.7 style hashing in Python 3.X+
 
 This is designed for compatibility not performance.
```
Comment (hoffa) [inline samtranslator/third_party/py27hash/hash.py:4]: 😏

File: samtranslator/third_party/py27hash/hash.py (hunk 2 above)
Comment (hoffa) [reply samtranslator/third_party/py27hash/hash.py:21]: Ah ok :( Even when you put `# type: ignore` on the exact line that it complains about?

File: samtranslator/third_party/py27hash/hash.py (hunk 3 above)
Comment (acristin) [reply samtranslator/third_party/py27hash/hash.py:4]: Yeah :/

**Review by hoffa — APPROVED** [review]

**Pull request merged**